### PR TITLE
low frequency filter

### PR DESCRIPTION
**Pull request by ccccjunkang**
## **Motivation**

Low-frequency features may occupy most of the memory, and it will also casue overfitting. Some training framework use a hashmap which counting features to filter low-frequency features, but it occupy some memeroy too.

We provide probabilistic-filter to solve this problem. When we encounter a feature that is not already in our model, we only add it to the model with probability p. Specifically, if a feature that is not already in our model, generate a random number.If the random number is less than the p, add the feature to the model.

filter_flow

reference: [https://dl.acm.org/doi/pdf/10.1145/2487575.2488200](https://dl.acm.org/doi/pdf/10.1145/2487575.2488200)

Example:

```Python
# filter_ratio is the probability p
v = sok.DynamicVariable(dimension=3,filter_ratio=0.1)
# open filter
embeddings = sok.lookup_sparse(
                [v], indices, combiners=combiners, use_low_frequency_filter=use_filter
)
```

## Modifed

1. Add an ratio_filter in hkv_variable(sparse_operation_kit/kit_src/variable/impl/hkv_variable.cu).In function ratio_filter, check if the key exist in hkv, by using hkv.find() which will return a bool vector(true mean the key already exists in the model). Then, if the key doesn't exist(determined by the bool vector),generate a random number(reuse curand state in hkv variable). If the random number is less than the p, set the key in bool vector to true(mean the key can be add in model).
2. Add an ratio_filter function in DummyVarAdapter, which call hkv_variable.ratio_filter. This function will merge ratio_filter request which from the same hkv variable.
3. Modifed sparse_forward_per_gpu. Use cub::DeviceSelect::Flagged to get the key being to lookup according to the bool vector as mentioned previously (true mean the key can be add in model or already in model). Counting the key after filter of each bucket and use cub::DeviceScan::InclusiveSum and to calculate the new bucket range.

Note:

This filter doesn't support det_variable and weighted_sparse_forward_per_gpu

## Testing

We used a business model and trained it with data for 3 months, and the AUC increased by one thousandth。The filter will not cost too much time.(about 1.5ms).In this case, per gpu has 2 million key,  after filtering there is 1.6 million key.

### CODE REVIEW

File: HugeCTR/embedding/all2all_embedding_collection.cu [hunk 1]
```diff
@@ -369,14 +369,88 @@ void weighted_sparse_forward_per_gpu(
...
+
+      key_t *keys_ptr = emb_input.keys.data<key_t>();
+
+      cub::DeviceSelect::Flagged(temp_select_storage.data(), temp_select_bytes, keys_ptr,
+                                 filterd_ptr, keys_after_filter.data<key_t>(),
+                                 emb_input.num_keys.data<uint64_t>(), emb_input.h_num_keys, stream);
+
+      size_t batch_size = (bucket_num) / meta.num_lookup_;
+
+      cal_lookup_idx<<<1, block_size, 0, stream>>>(meta.num_lookup_ + 1,
```
Comment (shijieliu) [inline HugeCTR/embedding/all2all_embedding_collection.cu:434]: why the grid_size is 1?

Comment (ccccjunkang) [reply HugeCTR/embedding/all2all_embedding_collection.cu:434]: This function is used to calculate the new lookup_offset, and the lookup_num is usually not particularly large. About a few hundred.
Let me describe the process of filter again. After enabling filter, a filtered boolean vector will be created, and then execute emb_storage ->ratio_filter. This function will check if the key exists in the map and randomly filter out some keys that do not exist. The filtered keys will be set to false in the filtered vector. Then, in the filter function, new keys, new bucket_range, and new lookup_offset will be calculated  according to the filtered vector.

Comment (kanghui0204) [reply HugeCTR/embedding/all2all_embedding_collection.cu:434]: @shijieliu do you think the response is OK ? and other part of @ccccjunkang code LGTM, if you think this response is OK , maybe we can merge it ?

**Comment by kanghui0204** [comment]
LGTM

File: HugeCTR/embedding/all2all_embedding_collection.cu (hunk 1 above)
Comment (shijieliu) [reply HugeCTR/embedding/all2all_embedding_collection.cu:434]: LGTM

**Comment by shijieliu** [comment]
LGTM.

**Pull request merged**